Mathlib/Data/Real/CauSeq.lean

```diff
@@ -0,0 +1,1017 @@
+applicable, lemmas that will be reused in other contexts have been stated in extra generality.
+There are other "versions" of Cauchyness in the library, in particular Cauchy filters in topology.
+This is a concrete implementation that is useful for simplicity and computability reasons.
+
+## Important definitions
+
+* `IsCauSeq`: a predicate that says `f : ℕ → β` is Cauchy.
+* `CauSeq`: the type of Cauchy sequences valued in type `β` with respect to an absolute value
+  function `abv`.
+
+## Tags
+
+sequence, cauchy, abs val, absolute value
+-/
+
+
+open IsAbsoluteValue
+
+theorem exists_forall_ge_and {α} [LinearOrder α] {P Q : α → Prop} :
+    (∃ i, ∀ j ≥ i, P j) → (∃ i, ∀ j ≥ i, Q j) → ∃ i, ∀ j ≥ i, P j ∧ Q j
+  | ⟨a, h₁⟩, ⟨b, h₂⟩ =>
+    let ⟨c, ac, bc⟩ := exists_ge_of_linear a b
+    ⟨c, fun _ hj => ⟨h₁ _ (le_trans ac hj), h₂ _ (le_trans bc hj)⟩⟩
+#align exists_forall_ge_and exists_forall_ge_and
+
+section
+
+variable [LinearOrderedField α] [Ring β] (abv : β → α) [IsAbsoluteValue abv]
+
+theorem rat_add_continuous_lemma {ε : α} (ε0 : 0 < ε) :
+    ∃ δ > 0,
+      ∀ {a₁ a₂ b₁ b₂ : β}, abv (a₁ - b₁) < δ → abv (a₂ - b₂) < δ → abv (a₁ + a₂ - (b₁ + b₂)) < ε :=
+  ⟨ε / 2, half_pos ε0, @fun a₁ a₂ b₁ b₂ h₁ h₂ => by
+    simpa [add_halves, sub_eq_add_neg, add_comm, add_left_comm, add_assoc] using
+      lt_of_le_of_lt (abv_add _ _) (add_lt_add h₁ h₂)⟩
+#align rat_add_continuous_lemma rat_add_continuous_lemma
+
+theorem rat_mul_continuous_lemma {ε K₁ K₂ : α} (ε0 : 0 < ε) :
+    ∃ δ > 0,
+      ∀ {a₁ a₂ b₁ b₂ : β},
+        abv a₁ < K₁ →
+          abv b₂ < K₂ → abv (a₁ - b₁) < δ → abv (a₂ - b₂) < δ → abv (a₁ * a₂ - b₁ * b₂) < ε :=
+  by
+  have K0 : (0 : α) < max 1 (max K₁ K₂) := lt_of_lt_of_le zero_lt_one (le_max_left _ _)
+  have εK := div_pos (half_pos ε0) K0
+  refine' ⟨_, εK, @fun a₁ a₂ b₁ b₂ ha₁ hb₂ h₁ h₂ => _⟩
+  replace ha₁ := lt_of_lt_of_le ha₁ (le_trans (le_max_left _ K₂) (le_max_right 1 _))
+  replace hb₂ := lt_of_lt_of_le hb₂ (le_trans (le_max_right K₁ _) (le_max_right 1 _))
+  have :=
+    add_lt_add (mul_lt_mul' (le_of_lt h₁) hb₂ (abv_nonneg _) εK)
+      (mul_lt_mul' (le_of_lt h₂) ha₁ (abv_nonneg _) εK)
+  rw [← abv_mul, mul_comm, div_mul_cancel _ (ne_of_gt K0), ← abv_mul, add_halves] at this
+  simpa [sub_eq_add_neg, mul_add, add_mul, add_left_comm] using lt_of_le_of_lt (abv_add _ _) this
+#align rat_mul_continuous_lemma rat_mul_continuous_lemma
+
+theorem rat_inv_continuous_lemma {β : Type _} [DivisionRing β] (abv : β → α) [IsAbsoluteValue abv]
+    {ε K : α} (ε0 : 0 < ε) (K0 : 0 < K) :
+    ∃ δ > 0, ∀ {a b : β}, K ≤ abv a → K ≤ abv b → abv (a - b) < δ → abv (a⁻¹ - b⁻¹) < ε := by
+  refine' ⟨K * ε * K, mul_pos (mul_pos K0 ε0) K0, @fun a b ha hb h => _⟩
+  have a0 := K0.trans_le ha
+  have b0 := K0.trans_le hb
+  rw [inv_sub_inv' ((abv_pos abv).1 a0) ((abv_pos abv).1 b0)]
+  rw [@abv_mul _ _ _ _ abv _ _ b⁻¹] -- Porting note: mistake in `abv_mul`?
```

Comment:
I think we should fix this now rather than working around it in every downstream port.

Reply:
Made a PR for that: #1398